File: DisappearedNumberArray.java
/*
 *  Time Complexity: O(N) where N is the length of the nums array.
 *  Space Complexity: O(1) Since we are mutating the existing array.
 *
 *  Did this code successfully run on Leetcode : Yes
 *  Any problem you faced while coding this : None
 *
 *  Explanation: For every value in the nums array we mark the number at that index  with a negative value. At the end all the indexes which are positive are the missing values since they did not appear in the nums array.
 */
class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        if(nums==null || nums.length==0) return new ArrayList<Integer>();
        List<Integer> list = new ArrayList<>();
        for(int i=0;i<nums.length;i++){
            if(nums[Math.abs(nums[i])-1]>0) nums[Math.abs(nums[i])-1]=nums[Math.abs(nums[i])-1]*-1;
        }
        for(int i=0;i<nums.length;i++){
            if(nums[i]>0) list.add(i+1);
        }
        return list;
    }
}

File: disappearednumberArray.py
# https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/
# Time O(N) : Two passes over nums of length N
# Space O(1) : No extra space used
# Can be done using hashset as extra space O(N) space and time.

class Solution:
    def findDisappearedNumbers(self, nums):
        # Time O(N) (2N) Space O(1)
        result = []
        #first pass mark the numbers we encounter as visted by making them negative
        for i in range(len(nums)):
            if nums[abs(nums[i]) - 1] > 0:
                nums[abs(nums[i])-1] *= -1
        
        #second pass to find all the non-negative indices for result
        for i in range(len(nums)):
            if nums[i] > 0:
                result.append(i+1)
                
        return result

nums = [4,3,2,7,8,2,3,1]
s = Solution()
output = s.findDisappearedNumbers(nums)
print(output)
        
        